pyomo/contrib/pynumero/sparse/block_matrix.py

```diff
@@ -341,385 +540,763 @@ def is_empty_block(self, idx, jdx):
 
-        if not isinstance(key, tuple):
-            raise RuntimeError('Wrong index')
+        assert row < self.bshape[0] and col < self.bshape[1], 'Indices out of range'
 
-        idx, jdx = key
-        assert idx >= 0 and jdx >= 0, 'indices must be positive'
-        assert idx < self.bshape[0] and jdx < self.bshape[1], 'indices out of range'
         if value is None:
-            self._blocks[idx, jdx] = None
-            self._block_mask[idx, jdx] = False
-            all_none_rows = True
-            for blk in self._blocks[:, jdx]:
-                if blk is not None:
-                    all_none_rows = False
-                    break
-
-            all_none_cols = True
-            for blk in self._blocks[idx, :]:
-                if blk is not None:
-                    all_none_cols = False
-                    break
-
-            if all_none_cols:
-                self._brow_lengths[idx] = 0
-            if all_none_rows:
-                self._bcol_lengths[jdx] = 0
+            self._blocks[row, col] = None
+            self._block_mask[row, col] = False
         else:
-            msg = 'blocks need to be sparse matrices'
-            assert isinstance(value, BlockMatrix) or isspmatrix(value), msg
-            if self._brow_lengths[idx] == 0 and self._bcol_lengths[jdx] == 0:
-                self._blocks[idx, jdx] = value
-                self._brow_lengths[idx] = value.shape[0]
-                self._bcol_lengths[jdx] = value.shape[1]
-                self._block_mask[idx, jdx] = True
-            elif self._brow_lengths[idx] != 0 and self._bcol_lengths[jdx] == 0:
-                if self._brow_lengths[idx] != value.shape[0]:
-                    msg = ('Incompatible row dimensions for block ({i},{j}) '
-                           'Got {got}, '
-                           'expected {exp}.'.format(i=idx, j=jdx,
-                                                    exp=self._brow_lengths[idx],
-                                                    got=value.shape[0]))
-                    raise RuntimeError(msg)
-                self._blocks[idx, jdx] = value
-                self._block_mask[idx, jdx] = True
-                self._bcol_lengths[jdx] = value.shape[1]
-            elif self._brow_lengths[idx] == 0 and self._bcol_lengths[jdx] != 0:
-                if self._bcol_lengths[jdx] != value.shape[1]:
-                    msg = ('Incompatible col dimensions for block ({i},{j}) '
-                           'Got {got}, '
-                           'expected {exp}.'.format(i=idx, j=jdx,
-                                                    exp=self._bcol_lengths[jdx],
-                                                    got=value.shape[1]))
-                    raise RuntimeError(msg)
-                self._blocks[idx, jdx] = value
-                self._block_mask[idx, jdx] = True
-                self._brow_lengths[idx] = value.shape[0]
+            if isinstance(value, BaseBlockMatrix):
```

Comment:
It will require quite a bit of work to support 2D NumPy arrays. For now, I think we should raise an exception if `set_block` gets a NumPy array and require the user to do the conversion. I just tested the dot product of a sparse 1x3 matrix with a sparse 3x1 matrix, and the result was a sparse 1x1 matrix. Can you give an example where this crops up?

Reply:
I guess the only concern I have with converting a numpy array to a Scipy sparse matrix under the hood is that we don't copy anything else the user puts into a BlockMatrix.

Reply:
@michaelbynum take a look at #1282 and change line 69 to `Z[0,0] = -inv(Ab)* Ad` I believe that is an example where that happens

Reply:
@santiagoropb Thanks. I will take a look.

Reply:
For now, can we copy the numpy array to a scipy sparse matrix and issue a warning?

Reply:
Yes that sounds reasonable

Reply:
Okay, I updated this.